USB Lab (#770)
UPB-PMRust/website: website/lab/09/index.mdx

use panic_probe as _;

// Randomly generated GUID to allow clients on Windows to find our device
const DEVICE_INTERFACE_GUIDS: &[&str] = &["{AFB9A6FB-30BA-44BC-9232-806CFC875321}"];

bind_interrupts!(struct Irqs {
    USB => InterruptHandler<USB>;
});
```

2. **Clock Configuration and Initialization**
USB requires a precise 48 MHz clock to function. On the STM32U5, we configure the internal HSI48 oscillator specifically for this purpose before initializing the peripherals.

```rust
#[embassy_executor::main]
async fn main(_spawner: Spawner) {
    info!("Hello");

    let mut config = Config::default();
    {
        use embassy_stm32::rcc::*;

        // Do not configure HSE or PLLs. Use internal oscillators.
        config.rcc.hsi = true;
        config.rcc.sys = Sysclk::HSI; // System clock is now 16MHz

        // Enable the 48MHz oscillator for USB
        config.rcc.hsi48 = Some(Hsi48Config {
            sync_from_usb: false, // Must be false
        });

        config.rcc.mux.iclksel = mux::Iclksel::HSI48;
        config.rcc.voltage_range = VoltageScale::RANGE2;
    }
    
    // Initialize peripherals with the custom clock config

>     let peripherals = embassy_stm32::init(config);

Reviewer: ```suggestion
    let peripherals = embassy_stm32::init(config);
    
    /* additional USB initialisation code follows here */

}
```

```

3. **USB Driver and Configuration Setup**
Next, we initialize the hardware-specific USB driver using the `PA12` (DP) and `PA11` (DM) pins. We also set up the `UsbConfig`, which defines the global properties of our device (like Vendor ID, Product ID, and manufacturer strings).

```rust
let driver = Driver::new(peripherals.USB, Irqs, peripherals.PA12, peripherals.PA11);

>     // Create embassy-usb Config

Reviewer: ```suggestion
    let driver = Driver::new(peripherals.USB, Irqs, peripherals.PA12, peripherals.PA11);

    // Create embassy-usb Config
```

    let mut config = UsbConfig::new(0xc0de, 0xcafe);
    config.manufacturer = Some("PMRust");
    config.product = Some("USB Bulk Example");
    config.serial_number = Some("0xcafe_c0de");
    config.max_power = 100;
    config.max_packet_size_0 = 64;
```

4. **Descriptor Builder and Windows Compatibility**
Embassy requires memory buffers to build the descriptor tree. We pass these to a `Builder`.
We also add **Microsoft OS Descriptors (MSOS)**. By telling Windows our device is compatible with the "WINUSB" feature, Windows will automatically load the standard WinUSB driver.

```rust
// Buffers for building the descriptors.

>     let mut config_descriptor = [0; 256];

Reviewer: ```suggestion
    // Buffers for building the descriptors.
    let mut config_descriptor = [0; 256];
```

    let mut bos_descriptor = [0; 256];
    let mut msos_descriptor = [0; 256];
    let mut control_buf = [0; 64];

    let mut builder = Builder::new(
        driver,
        config,
        &mut config_descriptor,
        &mut bos_descriptor,
        &mut msos_descriptor,
        &mut control_buf,
    );

    // Tell Windows to use the built-in WinUSB driver automatically
    builder.msos_descriptor(windows_version::WIN8_1, 0);
    builder.msos_feature(msos::CompatibleIdFeatureDescriptor::new("WINUSB", ""));
    builder.msos_feature(msos::RegistryPropertyFeatureDescriptor::new(
        "DeviceInterfaceGUIDs",
        msos::PropertyData::RegMultiSz(DEVICE_INTERFACE_GUIDS),
    ));
```

5. **Interface and Endpoint Creation**
We add a vendor-specific function (Class `0xFF`) and create our endpoints. We need two endpoints for our echo device: one Bulk OUT (host to device) and one Bulk IN (device to host).

```rust
// Add a vendor-specific function (class 0xFF) and interface

>     let mut function = builder.function(0xFF, 0, 0);

Reviewer: ```suggestion
    // Add a vendor-specific function (class 0xFF) and interface
    let mut function = builder.function(0xFF, 0, 0);
```

    let mut interface = function.interface();
    let mut alt = interface.alt_setting(0xFF, 0, 0, None);
    
    let mut read_ep = alt.endpoint_bulk_out(None, 64);
    let mut write_ep = alt.endpoint_bulk_in(None, 64);

>     drop(function);

Reviewer: Please explain what each of these lines of code does.

Author: Done.

    // Build the USB device handler
    let mut usb = builder.build();
```

6. **Running the Async Tasks**
USB communication is inherently asynchronous. We need to run the core USB driver task, alongside our application task that handles reading and writing.

Our read/write task waits for a connection, reads data from the `read_ep`, waits for 1 second, and writes that exact data back to the `write_ep`. Finally, we use join to run both tasks concurrently.

```rust
// Task 1: Run the core USB driver

>     let usb_driver = usb.run();

Reviewer: ```suggestion
    // Task 1: Initialize the core USB driver's task.
    let usb_driver = usb.run();
```

>     // Task 2: The actual application logic

Reviewer: ```suggestion
    // Task 2: Define the actual application logic
```

    let usb_read_write = async {
        loop {
            read_ep.wait_enabled().await;
            info!("Connected");
            
            loop {
                let mut data = [0; 64];
                match read_ep.read(&mut data).await {
                    Ok(n) => {
                        info!("Got bulk: {:a}", data[..n]);
                        Timer::after_secs(1).await;
                        
                        // Echo the data back to the host:
                        write_ep.write(&data[..n]).await.ok();
                    }
                    Err(_) => break, // Break the inner loop if disconnected
                }
            }
            info!("Disconnected");
        }
    };

>     // Run everything concurrently

Reviewer: ```suggestion
    // Run the USB driver task and the actual application logic task concurrently.
```

    join(usb_driver, usb_read_write).await;
}
```

## Exercises

:::warning 

> Before running any USB device code, you must configure your lab board correctly:

Reviewer: Can we  add a picture here?

Author: Yes, we should have a photo with the jumper here.

1. **Move Jumper JP3** from the `STLK` position to the `USB USER` position.

2. Connect **TWO USB** cables to your computer:

    - One to the `ST-LINK` port (for flashing the code and reading `defmt` logs).

    - One to the `USB USER` port (to connect your newly programmed USB device to the host).
:::

1. **Initialize the USB Peripheral (2p)**: Configure the USB builder on your microcontroller with a custom Vendor ID (`0xABCD`) and Product ID (`0x1234`). Set the manufacturer string to your name.
2. **Create a Bulk Echo Device (3p)**: Using the `embassy-usb` library, create an interface with two Bulk endpoints (one `IN` and one `OUT`). Implement an asynchronous loop that waits for data on the `OUT` endpoint, reverses the received byte array, and sends it back to the host via the `IN` endpoint.

> 3. **Host-side Rust Client (2p)**: Write a Rust script using nusb that:

Reviewer: ```suggestion
3. **Host-side Rust Client (2p)**: Write a Rust command line application using `nusb` that:
```

They should be able to use python if they want.

Author: Ok, that is good, but I think it will be better to make the exercises with the client in Rust in order the students to use the knowledge from the Lab0 - Intro to Rust. But I will add also the option to write the client in Python if you think that is a good idea.

> 

Reviewer: ```suggestion
```

    - Finds your custom device (VID=`0xABCD`, PID=`0x1234`).
    
    - Sends the string "PM USB Lab" to the OUT endpoint.
    
    - Reads the reversed string from the IN endpoint.
    

>     - Prints the result to the console.

Reviewer: ```suggestion
    - Finds your custom device (VID=`0xABCD`, PID=`0x1234`).
    - Sends the string "PM USB Lab" to the OUT endpoint.
    - Reads the reversed string from the IN endpoint.
    - Prints the result to the console.
```

---
description: Universal Serial Bus v2.0

> slug: /lab/09

Reviewer: Place the lab after the I2C and hide the others.

Author: Done.

    let driver = Driver::new(peripherals.USB, Irqs, peripherals.PA12, peripherals.PA11);

    // Create embassy-usb Config
    let mut config = UsbConfig::new(0xc0de, 0xcafe);
    config.manufacturer = Some("PMRust");
    config.product = Some("USB Bulk Example");
    config.serial_number = Some("0xcafe_c0de");
    config.max_power = 100;
    config.max_packet_size_0 = 64;
```
  </TabItem>

  <TabItem value="rp2350" label="Raspberry Pi Pico 2" default>

```rust
// Create the driver, from the HAL.
let driver = Driver::new(peripherals.USB, Irqs);

    // Create embassy-usb Config
    let mut config = UsbConfig::new(0xc0de, 0xcafe);
    config.manufacturer = Some("PMRust");
    config.product = Some("USB Bulk Example");
    config.serial_number = Some("0xcafe_c0de");
    config.max_power = 100;
    config.max_packet_size_0 = 64;
```

  </TabItem>

</Tabs>

4. **Descriptor Builder and Windows Compatibility**
Embassy requires memory buffers to build the descriptor tree. We pass these to a `Builder`.
We also add **Microsoft OS Descriptors (MSOS)**. By telling Windows our device is compatible with the "WINUSB" feature, Windows will automatically load the standard WinUSB driver.

> <Tabs>

Reviewer: Is this rp2040 code different?

Author: No.

    ));
```

  </TabItem>

  <TabItem value="rp2350" label="Raspberry Pi Pico 2" default>

```rust
    // Buffers for building the descriptors.
    let mut config_descriptor = [0; 256];
    let mut bos_descriptor = [0; 256];
    let mut msos_descriptor = [0; 256];
    let mut control_buf = [0; 64];
    
    let mut builder = Builder::new(
        driver,
        config,
        &mut config_descriptor,
        &mut bos_descriptor,
        &mut msos_descriptor,
        &mut control_buf,
    );
    
    // Tell Windows to use the built-in WinUSB driver automatically
    builder.msos_descriptor(windows_version::WIN8_1, 0);
    builder.msos_feature(msos::CompatibleIdFeatureDescriptor::new("WINUSB", ""));
    builder.msos_feature(msos::RegistryPropertyFeatureDescriptor::new(
        "DeviceInterfaceGUIDs",
        msos::PropertyData::RegMultiSz(DEVICE_INTERFACE_GUIDS),
    ));
```

  </TabItem>

</Tabs>

5. **Interface and Endpoint Creation**
We add a vendor-specific function (Class `0xFF`) and create our endpoints. We need two endpoints for our echo device: one Bulk OUT (host to device) and one Bulk IN (device to host).

<Tabs>

>   <TabItem value="stm32u5" label="STM32 Nucleo-U545RE-Q" default>

Reviewer: Is the rp2040 code different?

Author: No.
